File: src/comms/JsonGzipEncoder.py
from typing import Type
import gzip

from .JsonEncoder import JsonEncoder
from .jsonRpc import Request, Response, RequestOrResponse

class JsonGzipEncoder(JsonEncoder):

    def name(self) -> str:
        return 'json+gzip'

    def decode(self, message: bytes, cls: Type[RequestOrResponse]) -> RequestOrResponse:
        return JsonEncoder.decode(self, gzip.decompress(message), cls)

    def encode(self, obj: RequestOrResponse) -> bytes:
        return gzip.compress(JsonEncoder.encode(self, obj))

File: js/src/comms/JsonGzipEncoder.ts
import pako from 'pako'

import Encoder from './Encoder'
import JsonEncoder from './JsonEncoder'
import JsonRpcRequest from './JsonRpcRequest'
import JsonRpcResponse from './JsonRpcResponse'

/**
 * A [JSON](http://json.org/) + gzip message encoder
 */
export default class JsonGzipEncoder extends Encoder {

  /**
   * Used to perform the underlying JSON encodong/decoding.
   * Unable to extend `JsonEncoder` because this handles binary,
   * rather than string, messages.
   */
  private json: JsonEncoder

  constructor () {
    super()
    this.json = new JsonEncoder()
  }

  name (): string {
    return 'json+gzip'
  }

  encode (instance: JsonRpcRequest | JsonRpcResponse): Uint8Array {
    const json = this.json.encode(instance)
    return pako.deflate(json)
  }

  decode<Type = JsonRpcRequest | JsonRpcResponse> (message: Uint8Array, constructor: new () => Type): Type {
    const json = pako.inflate(message, { to: 'string' })
    return this.json.decode(json, constructor)
  }
}
